pvlib/tests/ivtools/test_sdm.py

```diff
@@ -0,0 +1,377 @@
+    # Failing to estimate the parameters for the CEC SDM model should raise an
+    # exception.
+    with pytest.raises(RuntimeError):
+        sdm.fit_cec_sam(celltype='polySi', v_mp=0.45, i_mp=5.25, v_oc=0.55,
+                        i_sc=5.5, alpha_sc=0.00275, beta_voc=0.00275,
+                        gamma_pmp=0.0055, cells_in_series=1, temp_ref=25)
+
+
+@requires_scipy
+def test_fit_desoto():
+    result, _ = sdm.fit_desoto(v_mp=31.0, i_mp=8.71, v_oc=38.3, i_sc=9.43,
+                               alpha_sc=0.005658, beta_voc=-0.13788,
+                               cells_in_series=60)
+    result_expected = {'I_L_ref': 9.45232,
+                       'I_o_ref': 3.22460e-10,
+                       'R_s': 0.297814,
+                       'R_sh_ref': 125.798,
+                       'a_ref': 1.59128,
+                       'alpha_sc': 0.005658,
+                       'EgRef': 1.121,
+                       'dEgdT': -0.0002677,
+                       'irrad_ref': 1000,
+                       'temp_ref': 25}
+    assert np.allclose(pd.Series(result), pd.Series(result_expected),
+                       rtol=1e-4)
+
+
+@requires_scipy
+def test_fit_desoto_failure():
+    with pytest.raises(RuntimeError) as exc:
+        sdm.fit_desoto(v_mp=31.0, i_mp=8.71, v_oc=38.3, i_sc=9.43,
+                       alpha_sc=0.005658, beta_voc=-0.13788,
+                       cells_in_series=10)
+    assert ('Parameter estimation failed') in str(exc.value)
+
+
+@requires_scipy
+@requires_statsmodels
+def test_fit_desoto_sandia(cec_params_cansol_cs5p_220p):
+    # this test computes a set of IV curves for the input fixture, fits
+    # the De Soto model to the calculated IV curves, and compares the fitted
+    # parameters to the starting values
+    params = cec_params_cansol_cs5p_220p['params']
+    params.pop('Adjust')
+    specs = cec_params_cansol_cs5p_220p['specs']
+    effective_irradiance = np.array([400., 500., 600., 700., 800., 900.,
+                                     1000.])
+    temp_cell = np.array([15., 25., 35., 45.])
+    ee = np.tile(effective_irradiance, len(temp_cell))
+    tc = np.repeat(temp_cell, len(effective_irradiance))
+    iph, io, rs, rsh, nnsvth = pvsystem.calcparams_desoto(
+        ee, tc, alpha_sc=specs['alpha_sc'], **params)
+    sim_ivcurves = pvsystem.singlediode(iph, io, rs, rsh, nnsvth, 300)
+    sim_ivcurves['ee'] = ee
+    sim_ivcurves['tc'] = tc
+
+    result = sdm.fit_desoto_sandia(sim_ivcurves, specs)
+    modeled = pd.Series(index=params.keys(), data=np.nan)
+    modeled['a_ref'] = result['a_ref']
+    modeled['I_L_ref'] = result['I_L_ref']
+    modeled['I_o_ref'] = result['I_o_ref']
+    modeled['R_s'] = result['R_s']
+    modeled['R_sh_ref'] = result['R_sh_ref']
+    expected = pd.Series(params)
+    assert np.allclose(modeled[params.keys()].values,
+                       expected[params.keys()].values, rtol=5e-2)
+
+
+@requires_scipy
+@requires_statsmodels
+def test_fit_pvsyst_sandia(npts=3000):
```

Comment:
can this test be broken up into much smaller chunks?

Reply:
it's a long test function but I don't see the advantage of breaking it up. There's only one test with many outputs being checked.

The test produces 1) Pvsyst parameters (9 values), 2) five diode equation values for each IV curve, and 3) a Boolean array which is True if the parameter procedure converged for an IV curve.  Each of these outputs is tested. Also, the estimated Pvsyst parameters are pushed forward through the Pvsyst model and the calculated IV curves are compared with the input IV curves.

Half of the code is extracting variables from the text files containing the input IV curves and the known solution.  We could separate the IO portion into a helper function but it would only be used here.

Long term, if pvlib adopts an explicit convention for IV curve data, there would be code in pvlib.iotools to read/write that data, and the text files used here would conform to that convention.

Reply:
I think a helper function or two would make the test more readable but fine to leave as is.